File: examples/acquisition_model_and_data_algebra.py
'''
Upper-level demo that illustrates the computation of coil sensitivity maps,
applying projection from the image space into acquisition space and back
defined by the aquisition model, and images and acquisitions algebra.
'''

import argparse
import os
import sys

BUILD_PATH = os.environ.get('BUILD_PATH') + '/xGadgetron'
SRC_PATH = os.environ.get('SRC_PATH') + '/xGadgetron/pGadgetron'

sys.path.append(BUILD_PATH)
sys.path.append(SRC_PATH)

from pGadgetron import *

parser = argparse.ArgumentParser(description = \
'''
Upper-level demo that illustrates the computation of coil sensitivity maps,
applying projection from the image space into acquisition space and back
defined by the aquisition model, and images and acquisitions algebra.
''')
parser.add_argument\
('filename', nargs='?', default = 'testdata.h5', \
 help = 'raw data file name (default: testdata.h5)')
args = parser.parse_args()                                 

def main():

    # acquisitions will be read from an HDF file args.filename
    input_data = MR_Acquisitions(args.filename)

    print('---\n acquisition data norm: %e' % input_data.norm())

    print('---\n processing acquisitions...')
    processed_data = input_data.process(['RemoveROOversamplingGadget'])

    print('---\n processed acquisition data norm: %e' % processed_data.norm())

    # perform reconstruction
    recon = MR_BasicReconstruction()
    recon.set_input(processed_data)
    recon.process()
    complex_images = recon.get_output()

    print('---\n reconstructed images norm: %e' % complex_images.norm())

    csms = MR_CoilSensitivityMaps()

    print('---\n sorting acquisitions...')
    processed_data.sort()
    print('---\n computing sensitivity maps...')
    csms.calculate(processed_data)

    # create acquisition model based on the acquisition parameters
    # stored in input_data and image parameters stored in complex_images
    am = MR_AcquisitionModel(processed_data, complex_images)

    am.set_coil_sensitivity_maps(csms)

    # use the acquisition model (forward projection) to produce 'acquisitions'
    acqs = am.forward(complex_images)

    print('---\n reconstructed images forward projection norm %e' % acqs.norm())

    # compute the difference between real and modelled acquisitions
    diff = acqs - processed_data
    rr = diff.norm()/acqs.norm()
    print('---\n reconstruction residual norm (rel): %e' % rr)

    # compare reconstructed images with backward-projected acquisitions
    bwd_images = am.backward(processed_data)
    im_diff = bwd_images - complex_images
    print(\
        '---\n difference between reconstructed and back-projected images: %e'\
        % (im_diff.norm()/complex_images.norm()))

    # test that the backward projection is the adjoint of forward
    # on x = processed_data and y = complex_images
    xFy = processed_data * acqs
    Bxy = bwd_images * complex_images
    print('---\n (x, F y) = (%e, %e)' % (xFy.real, xFy.imag))
    print('= (B x, y) = (%e, %e)' % (Bxy.real, Bxy.imag))

    complex_images.show()

try:
    main()
    print('done')

except error as err:
    # display error information
    print('??? %s' % err.value)

File: examples/acquisition_model_and_data_algebra.m
% Upper-level demo, illustrates the use of acquisition models and images and acquisitions algebra.
% Involves the computation of coil sensitivity maps and
% the projection from the image space into acquisition space and back.

set_up_mr

try
    % acquisitions will be read from this HDF file
    file = input('raw data file: ', 's');
    input_data = AcquisitionData(file);
    fprintf('%d acquisitions found\n', input_data.number())
    
    % pre-process acquisition data
    fprintf('processing acquisitions...\n')
    processed_data = preprocess_acquisitions(input_data);
	
    % perform reconstruction
    recon = SimpleReconstruction();
    recon.set_input(processed_data)
    fprintf('reconstructing...\n')
    recon.process()
    complex_images = recon.get_output();
    
    csms = CoilSensitivityMaps();
    fprintf('sorting acquisitions...\n')
    processed_data.sort()
    fprintf('calculating sensitivity maps...\n')
    csms.calculate(processed_data)

    % create acquisition model based on the acquisition parameters
    % stored in input_data and image parameters stored in interim_images
    am = AcquisitionModel(processed_data, complex_images);

    am.set_coil_sensitivity_maps(csms)

    % use the acquisition model (forward projection) to produce acquisitions
    acqs = am.forward(complex_images);

    % compute the difference between real and modelled acquisitions
    diff = acqs - processed_data;
    fprintf('reconstruction residual: %e\n', diff.norm()/acqs.norm())

    % apply the adjoint model (backward projection)
    imgs = am.backward(processed_data);

    % test that the backward projection is the adjoint of forward
    % on x = diff and y = complex_images
    fprintf('(x, F y) = %s\n', num2str(processed_data * acqs))
    fprintf('= (B x, y) = %s\n', num2str(imgs * complex_images))

    % test images norm and dot product
    s = imgs.norm();
    fprintf('(B x, B x) = %s = %e\n', num2str(imgs * imgs), s*s)

    % test linear combination of images
    im_diff = imgs - complex_images;
    fprintf('0.0 = %e\n', im_diff.norm()/complex_images.norm())

    complex_images.show()

catch err
    % display error information
    fprintf('%s\n', err.message)
    fprintf('error id is %s\n', err.identifier)
end
